package pyqtc.pb;

message Message {
  optional int32 id = 1;

  optional ErrorResponse error_response = 2;

  optional CreateProjectRequest create_project_request = 3;
  optional CreateProjectResponse create_project_response = 4;

  optional DestroyProjectRequest destroy_project_request = 5;
  optional DestroyProjectResponse destroy_project_response = 6;

  optional CompletionRequest completion_request = 7;
  optional CompletionResponse completion_response = 8;

  optional TooltipRequest tooltip_request = 9;
  optional TooltipResponse tooltip_response = 10;

  optional DefinitionLocationRequest definition_location_request = 11;
  optional DefinitionLocationResponse definition_location_response = 12;

  optional RebuildSymbolIndexRequest rebuild_symbol_index_request = 13;
  optional RebuildSymbolIndexResponse rebuild_symbol_index_response = 14;

  optional UpdateSymbolIndexRequest update_symbol_index_request = 15;
  optional UpdateSymbolIndexResponse update_symbol_index_response = 16;

  optional SearchRequest search_request = 17;
  optional SearchResponse search_response = 18;
}

service WorkerService {
  rpc CreateProject (CreateProjectRequest) returns (CreateProjectResponse);
}

message Context {
  optional string file_path = 1;
  optional string source_text = 2;
  optional int32 cursor_position = 3;
}

message ErrorResponse {
  optional string message = 1;
}

message CreateProjectRequest {
  optional string project_root = 1;
}

message CreateProjectResponse {
}

message DestroyProjectRequest {
  optional string project_root = 1;
}

message DestroyProjectResponse {
  optional string project_root = 1;
}

message CompletionRequest {
  optional Context context = 1;
}

message CompletionResponse {
  message Proposal {
    enum Type {
      INSTANCE = 1;
      CLASS = 2;
      FUNCTION = 3;
      MODULE = 4;
    }

    enum Scope {
      LOCAL = 1;
      GLOBAL = 2;
      BUILTIN = 3;
      ATTRIBUTE = 4;
      IMPORTED = 5;
      KEYWORD = 6;
      PARAMETER_KEYWORD = 7;
    }

    optional string name = 1;
    optional Type type = 2;
    optional Scope scope = 3;
    optional string docstring = 4;
  }

  repeated Proposal proposal = 1;
  optional int32 insertion_position = 2;

  optional string calltip = 3;
}

message TooltipRequest {
  optional Context context = 1;
}

message TooltipResponse {
  optional string rich_text = 1;
}

message DefinitionLocationRequest {
  optional Context context = 1;
}

message DefinitionLocationResponse {
  optional string file_path = 1;
  optional int32 line = 2;
}

message RebuildSymbolIndexRequest {
  optional string project_root = 1;
}

message RebuildSymbolIndexResponse {
}

message UpdateSymbolIndexRequest {
  optional string file_path = 1;
}

message UpdateSymbolIndexResponse {
}

enum SymbolType {
  ALL = 0;
  MODULE = 1;
  CLASS = 2;
  FUNCTION = 3;
  VARIABLE = 4;
}

message SearchRequest {
  optional string query = 1;
  optional string file_path = 2;
  optional SymbolType symbol_type = 3;
}

message SearchResponse {
  message Result {
    optional string module_name = 1;
    optional string file_path = 2;
    optional int32 line_number = 3;
    optional string symbol_name = 4;
    optional SymbolType symbol_type = 5;
  }

  repeated Result result = 1;
}
